### PR TITLE
Fix taping for subfunctions

### LINKED ISSUE #2811: Trouble with vertex only mesh interpolation and adjoint
The following script can use either split(), .split(), or .subfunctions to split by uncommenting appropriately.
Two of these produce an error message, whilst .split() provokes the deprecation warning.

```
from firedrake import *
from firedrake_adjoint import *
import numpy as np

mesh = UnitSquareMesh(3,3)

V = FunctionSpace(mesh, "CG", 1)
W = V * V

w = Function(W)

u, v = w.subfunctions

u.assign(3.)
v.assign(3.)

coords = np.array([[0.5, 0.6],
                   [0.3, 0.4]])
VOM = VertexOnlyMesh(mesh, coords)
VVOM = FunctionSpace(VOM, "DG", 0)
vv = Function(VVOM)

#u, v = w.subfunctions
#u, v = w.split()
u, v = split(w)
vv.interpolate(u)

J = assemble(vv**2*dx)

Jhat = ReducedFunctional(J, Control(w))
Jhat.derivative()
```

### PR DESCRIPTION
**Pull request by connorjward**
Fixes #2811 

It would be good to add a test of this. I don't quite know what that would look like.

### CODE REVIEW

**Comment by ReubenHill** [comment]
Well it looks like two tests of split annotation now fail - can we keep both split and subfunctions annotation working?

**Comment by connorjward** [comment]
> Well it looks like two tests of split annotation now fail - can we keep both split and subfunctions annotation working unless and until split is deprecated?

Yes that's a good idea. Since we don't tape nested annotations we could keep the decorator on `split`. Will do.

File: firedrake/function.py [hunk 1]
```diff
@@ -304,14 +304,14 @@ def __dir__(self):
...
+    @FunctionMixin._ad_annotate_subfunctions
     def subfunctions(self):
         r"""Extract any sub :class:`Function`\s defined on the component spaces
         of this this :class:`Function`'s :class:`.FunctionSpace`."""
         return tuple(type(self)(V, val)
                      for (V, val) in zip(self.function_space(), self.topological.subfunctions))
 
-    @PETSc.Log.EventDecorator()
-    @FunctionMixin._ad_annotate_split
+    @FunctionMixin._ad_annotate_subfunctions
```
[the remark this answers is not in the document]
Comment (connorjward) [reply firedrake/function.py:314]: When we annotate a function it is modified to accept a few new kwargs. pyadjoint tests were failing because not decorating the deprecated function meant that these additional kwargs weren't recognised.

I should probably revisit the `zero` one to pass `args` and `kwargs` through to `assign`.

[the remark this answers is not in the document]
Comment (ReubenHill) [reply firedrake/function.py:314]: I.e. it didn't know about `.subfunctions` and also doesn't know about `.zero`?

[the remark this answers is not in the document]
Comment (connorjward) [reply firedrake/function.py:314]: pyadjoint is still using the deprecated `split` so yeah it doesn't know about `subfunctions`. Adding a pyadjoint test for `zero` is now on my TODO list.

**Review by ReubenHill — APPROVED** [review]

**Comment by ReubenHill** [comment]
Just before merging, does this stop #2811 producing an error?

**Comment by connorjward** [comment]
> Just before merging, does this stop #2811 producing an error?

Yes
